cpp/serve/model.h

```diff
@@ -20,15 +22,110 @@ using tvm::Device;
 using namespace tvm::runtime;
 
 /*!
- * \brief Create the runtime module for LLM functions.
- * \param reload_lib The model library. It might be a path to the binary
- * file or an executable module that is pre-loaded.
- * \param model_path The path to the model weight parameters.
- * \param device The device to run the model on.
- * \return The created runtime module.
+ * \brief The model module for LLM functions.
+ * It runs an LLM, and has an internal KV cache that maintains
+ * the history KV values of all processed tokens.
+ *
+ * It contains the following functions:
+ *
+ * Model related:
+ * - "token_embed": take token ids as input and return the embeddings,
+ * - "batch_prefill": take embedding of a single sequence
+ * as input, forward the embedding through LLM and return the logits,
+ * - "decode": take the embeddings of the last-committed token of an
+ * entire batch as input, forward through LLM and return the logits
+ * for all sequences in the batch,
+ * - "softmax_with_temperature": take logits and temperatures, return
+ * probabilities.
+ *
+ * KV cache related:
+ * - "create_kv_cache": create the KV cache for this module,
+ * - "add_new_sequence": add (declare) a new sequence in the KV cache,
+ * - "remove_sequence": remove a sequence from KV cache.
+ *
+ * ... and some other auxiliary functions.
  */
-MLC_LLM_DLL tvm::runtime::Module CreateModelModule(TVMArgValue reload_lib, String model_path,
-                                                   DLDevice device);
+class Model : public Object {
+ public:
+  /*!
+   * \brief Create the runtime module for LLM functions.
+   * \param reload_lib The model library. It might be a path to the binary
+   * file or an executable module that is pre-loaded.
+   * \param model_path The path to the model weight parameters.
+   * \param device The device to run the model on.
+   * \return The created runtime module.
+   */
+  TVM_DLL static ObjectPtr<Model> Create(TVMArgValue reload_lib, String model_path,
+                                         DLDevice device);
+
+  /*********************** Model Computation  ***********************/
+
+  /*!
+   * \brief Compute embeddings for the input token ids.
+   * \param token_ids The token ids to compute embedding for.
+   * \return The computed embeddings.
+   */
+  virtual NDArray TokenEmbed(IntTuple batch_token_ids) = 0;
+
+  /*!
+   * \brief Batch prefill function. Embedding in, logits out.
+   * \param embeddings The embedding of the input to be prefilled.
+   * \param seq_id The id of the sequence in the KV cache.
+   * \param lengths The length of each sequence to prefill.
+   * \return The logits for the next token.
+   */
+  virtual NDArray BatchPrefill(Array<NDArray> embedding_arr, std::vector<int> seq_ids,
```

Comment:
I'm trying to understand the logic here - `Array<NDArray>` of `embedding_arr` seems to indicate the underlying implementation of a Model needs to take care of NDArray concatenation? Did I get it right?

Reply:
Yes you get it right. Right now the `ModelImpl` class concatenates the input embeddings before passing to the `prefill_with_embed` function in VM. The concatenation happens for the sequence dimension.

My concern for this place is that, the concatenation (implemented in a series of cudaMemcpy) might have some extra overhead. For two reasons I still chose to go with the current interface:
1. During profiling I found this overhead of memcpy is negligible.
2. An alternative interface is directly taking the concatenated embeddings as input here. To concatenate outside the BatchPrefill function, we may need to support `TokenEmbed` to take a batch of token ids, and somehow let the VM `embed` function create a large NDArray first and then directly write the embeddings of different sequences into their respective target locations. It might be indeed faster, while this approach seems not as straightforward when we have more input modalities (such as a mixture of `TokenData` and `EmbeddingData`).

So from my POV, as long as the memcpy does not have obvious overhead, we can take a batch of embeddings for BatchPrefill and concatenate them internally. Does my elaboration here answer your question and make it a bit clearer?

Reply:
I was previously eyeing the approach 2 as you mentioned, but didn’t quite think through. Will take a deeper look tomorrow!